Example: News Classification with GPT-2 (#24)
opst/knitfab-docs: 04.examples/news-classification/news-classification.en.md

```bash
git clone https://github.com/opst/knitfab-docs.git
```
Once cloned, navigate to the `04.examples/news-classification` directory. You will find the following:
- **scripts:** This directory contains the essential files for training and deploying the model:
  - Python scripts (`train.py`, `evaluate.py`) for model training and evaluation.
  - `requirements.txt` files listing all required Python packages for dependency management.
  - `Dockerfile` used to build a Docker image of the application, enabling consistent deployment within the Knitfab Plan.
- **configs:** Contains the configuration for training and evaluation task.
- **plans:** Contains the Knitfab Plan YAML templates.

### Task
- [Step 1: Define the training and evaluation tasks.](#step-1-define-the-training-and-evaluation-tasks)
- [Step 2: Build and push docker image to Knitfab.](#step-2-build-and-push-docker-image-to-knitfab)
- [Step 3: Fine-tuning.](#step-3-fine-tuning)
- [Step 4: LLM-as-a-judge evaluation.](#step-4-llm-as-a-judge-evaluation)
- [Step 5: Clean up.](#step-5-clean-up)

## Step 1: Define the training and evaluation tasks

> ### 1-1. Define Training Task

Reviewer: ```suggestion
### 1.1. Define Training Task

```

Author: Using three-number headings (e.g., 1.1.1, 1.2.4) can create inconsistencies in the steps, especially when some steps have sub-steps and others don't (e.g., step 1 has 1.1.1 - 1.2.4, but step 4 has no sub-steps, appearing as just 4.1- 4.9). This can be confusing.
To improve clarity, I'll use heading levels (### and ####) instead of numbered sub-headings (like 1.1 or 1.2). Hence, 1.1 will be removed

This section provides an in-depth explanation of the logic implemented in `scripts/train/train.py` for fine-tuning a news classification model. The script follows a structured pipeline that includes argument parsing, model preparation, dataset processing, training, and evaluation. Each component is designed to facilitate easy customization and optimization.

> ### 1. Parse Arguments

Reviewer: ```suggestion
### 1.1.1. Parse Arguments

```

Reviewer: Numberring shoule be consistent and each section numbers shoule be unique in document.

Author: Using three-number headings (e.g., 1.1.1, 1.2.4) can create inconsistencies in the steps, especially when some steps have sub-steps and others don't (e.g., step 1 has 1.1.1 - 1.2.4, but step 4 has no sub-steps, appearing as just 4.1- 4.9). This can be confusing.  
To improve clarity, I'll use heading levels (### and ####) instead of numbered sub-headings (like 1.1 or 1.2). This will make the structure easier to understand.

#### Key Functionalities
- `--config-file`: Path to a configuration JSON file. If provided, settings in this file will override the command-line arguments.
- `--save-to`: Directory where the trained model and logs will be stored. Defaults to `./out`.
- `--base-model`: Specifies the name of the base model to use for fine-tuning. Defaults to `openai-community/gpt2`.
- `--device`: Defines the hardware on which the training will run (`cuda` or `cpu`). Defaults to `auto`.
- `--epochs`: Number of training epochs. Defaults to `1`.
- `--learning-rate`: Specifies the learning rate for model training. Defaults to `2e-5`.

If a configuration file is provided via `--config-file`, its settings will override the corresponding command-line arguments to ensure consistency across different training runs.

#### Code Breakdown
```python
args = parser.parse_args()
if args.config_file:
    with open(args.config_file, "r") as f:
        config = json.load(f)
    for key, value in config.items():
        setattr(args, key.replace("-", "_"), value)
```
- Reads configuration from a JSON file if provided.
- Updates argument values dynamically.

> ### 2. Initialize Model and Tokenizer

Reviewer: ```suggestion
### 1.1.2. Initialize Model and Tokenizer

```

...and so on

Author: Same as above.

```docker
# Modify the command (e.g., base-url, api-key) to set your LLM as Evaluator
RUN deepeval set-local-model --model-name=llama3.2 \
    --base-url="http://ollama:11434/v1/" \
    --api-key="ollama"
```
#### Build Docker Image:
```bash
docker build -t news-classification-evaluate:v1.0 \
             -f scripts/evaluate/Dockerfile \
             ./scripts/evaluate
```
The `news-classification-evaluate` image is used to evaluate the performance of the fine-tuned model with LLM-as-a-judge.

### 2-2. (Optional) To Verify Docker Images:
> [!Note]
> 
> If you're confident in your images, feel free to skip ahead to pushing them to Knitfab. ([To Push Docker Images to Knitfab](#2-3-to-push-docker-images-to-knitfab))

### 1. Run the Fine-tune Image:
```bash
docker run --rm -it --gpus all \
    -v "$(pwd)/configs:/configs" \
    -v "$(pwd)/out:/out" \
    news-classification-train:v1.0
```
This command runs the `news-classification-train:v1.0` image in an interactive mode.
- The `-v` flags mount the host directories containing the config files (`/configs`) and the output directory (`/out`) into the container.

> - This allows you to test the image locally and generate the fine-tuned model.

Reviewer: Suggest that user can stop the container if it seems running fine.
User does not need to *complete* the fine-tune task on Docker, I think

```
This command generates a YAML template based on the Docker image `news-classification-train:v1.0`.

\* Replace `${registry_uri}` with the actual URI of your Knitfab registry.

\* This approach can help automate some of the configuration process.

### 2. Modify YAML Template:
- Crucial Modifications:
  - `image`: 
    If your Knitfab Kubernetes Cluster utilizes a local registry, replace `registry_uri` within the `image` field with `localhost`. 

    Example:
    ```YAML
    # Replace 192.0.2.1
    image: "192.0.2.1:30503/news-classification-train:v1.0"
    # With localhost
    image: "localhost:30503/news-classification-train:v1.0"
    ```
    - This ensures the image is pulled from the local registry.

  - `inputs`, `outputs`, `log`: 
    Add the following tags to the `inputs`, `outputs` and `log` sections:

    ```YAML
    inputs:
      - path: "/configs"
        tags:
          - "type:config"
          - "project:news-classification"
    
    outputs:
      - path: "/out/model"
        tags:
          - "type:model"
          - "project:news-classification"
    
    log:
      tags:
        - "type:log"
        - "project:news-classification"
    ```

>     - This facilitates better organization and filtering of `inputs`, `outputs` and `log` within the Knitfab system.

Reviewer: And, this is requried to invoke following task by Knitfab.

  - resources:
    Define the required CPU, Memory, and GPU resources

    ```YAML
    resources:
      cpu: 1
      memory: 10Gi
      nvidia.com/gpu: "1"
    ```
    

- Other Important Considerations:
  Always double-check that your modified YAML template adheres to the correct structure and syntax. You can use the YAML file provided in the `/plans` directory of the cloned Git repository as a reference.

### 3. Apply the YAML Template:
```bash
train_plan=$(knit plan apply ./plans/news-classification-train.v1.0.yaml)
```

> This command sends the YAML template to the Knitfab API, which creates a new Plan based on the provided configuration. 

Reviewer: ```suggestion
This command sends the Plan YAML to the Knitfab API, which creates a new Plan based on the provided configuration. 

```

Hereby, the yaml file is filled and fixed, so not a "template".

docker save ${registry_uri}/news-classification-evaluate:v1.0 | \
    knit plan template > ./plans/news-classification-evaluate.v1.0.yaml
```
\* Replace `${registry_uri}` with the actual URI of your Knitfab registry.

### 2. Modify YAML Template:
- Crucial Modifications:
  - `image`: 
    If your Knitfab Kubernetes Cluster utilizes a local registry, replace `registry_uri` within the image field with `localhost`.

    Example:
    ```YAML
    # Replace 192.0.2.1
    image: "192.0.2.1:30503/news-classification-evaluate:v1.0"
    # With localhost
    image: "localhost:30503/news-classification-evaluate:v1.0"
    ```
  - `inputs`, `outputs`, `log`: 
    Add the following tags to the `inputs`, `outputs` and `log` sections:
    
    ```YAML
    inputs:
      - path: "/configs"
        tags:
          - "type:config"
          - "project:news-classification"
      
      - path: "/in"
        tags:
          - "type:model"
          - "project:news-classification"

    outputs:
      - path: "/out"
        tags:
          - "type:metrics"
          - "project:news-classification"

    log:
      tags:
        - "type:log"
        - "project:news-classification"
    ```
> [!Note]
> 

> > To ensure the evaluation dynamically recognizes the output model from the training process, we will *not* define the `in/model` within the configuration file. Instead, we will explicitly define `in/model` within the Plan YAML template.

Reviewer: ```suggestion
> To ensure the evaluation dynamically recognizes the output model from the training process, we will *not* define the `in/model` within the configuration file. Instead, we will explicitly define `in/model` within the Plan YAML.

```

To be clear, Knitfab Plan's `inputs.path` means just "mount the input Data here".
So, this has no implications about options of running command in the container.

This is absolute path, as it is, and not relative from current directory of the entrypoint of the container.

  - `requirements.txt` files listing all required Python packages for dependency management.
  - `Dockerfile` used to build a Docker image of the application, enabling consistent deployment within the Knitfab Plan.
- **configs:** Contains the configuration for training and evaluation task.
- **plans:** Contains the Knitfab Plan YAML templates.

### Task
- [Step 1: Define the training and evaluation tasks.](#step-1-define-the-training-and-evaluation-tasks)
- [Step 2: Build and push docker image to Knitfab.](#step-2-build-and-push-docker-image-to-knitfab)
- [Step 3: Fine-tuning.](#step-3-fine-tuning)
- [Step 4: LLM-as-a-judge evaluation.](#step-4-llm-as-a-judge-evaluation)
- [Step 5: Clean up.](#step-5-clean-up)

## Step 1: Define the training and evaluation tasks

### Define Training Task:

This section provides an in-depth explanation of the logic implemented in `scripts/train/train.py` for fine-tuning a news classification model. The script follows a structured pipeline that includes argument parsing, model preparation, dataset processing, training, and evaluation. Each component is designed to facilitate easy customization and optimization.

> #### 1. Parse Arguments

Reviewer: ```suggestion
#### 1.1. Parse Arguments

```

Section number is TO BE UNIQUE in document.

The number depth tells reader how overview or detailed contents are placed there.
And, users can point section with the unique number without confusion.

### Example
# Fine-Tune LLM model for News Classification with Knitfab

This guide walks you through fine-tuning a large language model (LLM) for classification using Knitfab. Discover how Knitfab automates the fine-tuning process and how you can leverage the LLM-as-a-judge approach for efficient, human-like evaluation.

### Overview
Fine-tuning LLMs involves optimizing multiple parameters to achieve optimal performance for a specific task—a process that can quickly become overwhelming. Knitfab provides robust management of training artifacts, facilitating seamless experimentation with different configurations. Additionally, we'll explore LLM-as-a-judge, an innovative evaluation method that delivers human-like assessment quality while reducing costs and time.

The following diagram outlines the key components and steps involved:

**Components:**

> - __Base Model:__ The pre-trained LLM used as a foundation for fine-tuning. In this example, we'll use GPT-2.

Reviewer: A bit broken.

```suggestion
- __Base Model:__ The pre-trained LLM used as a foundation for fine-tuning. In this example, we will use GPT-2.

```

- __Training Dataset:__ The dataset used to fine-tune the model for news classification. We will leverage the 20 Newsgroups dataset from scikit-learn.
- __Fine-tuned Model:__ The resulting model after fine-tuning, now specialized for news classification.
- __Evaluation Dataset:__ A held-out subset of the 20 Newsgroups dataset used for LLM-as-a-judge evaluation.
- __Metrics:__ Custom metrics designed to evaluate LLM outputs.

**Process:**
- __Train:__ Fine-tune the base LLM on the training dataset to classify news articles.
- __Evaluate:__ Assess the performance of the fine-tuned model using the defined custom metrics and the evaluation dataset.

```mermaid
flowchart LR
    A((Base Model)) --> C[Train]
    B((Training Dataset)) --> C
    C --> D((Fine-tuned Model))
    D --> F[Evaluate]
    E((Evaluation Dataset)) --> F
    F --> G((Metrics)) 

> ```

Reviewer: Comment here about deploying Ollama as plan Kubernetes resources(= outside of Knitfab), for LLM-as-a-judge.

User needs to access to bare Kubernetes cluster in this example, so give them a cue.

To successfully complete this example, ensure you have met the following prerequisites:

**Essential:**
- **GPU:** This example is designed for GPU execution.  Running it on a CPU is inefficient and will result in significantly longer processing times.
- **Knitfab Setup:** 
  - **Production:** Follow the instructions provided in [03.admin-guide](../../03.admin-guide/admin-guide-installation.en.md) for full installation.
  - **Experimentation:** Follow the instructions provided in [01.getting-started: Installing Knitfab on a local environment](../../01.getting-started/getting-started.en.md#installing-knitfab-on-a-local-environment) for a light weight version.
- **Knit CLI:** Follow the installation instructions in [01.getting-started: CLI Tool: knit](../../01.getting-started/getting-started.en.md#cli-tool-knit) to set up the Knit CLI.
- **Knit Command Initialization:** Refer to [01.getting-started: Initializing Knit Command](../../01.getting-started/getting-started.en.md#initializing-the-knit-command) for guidance on initializing Knit command using `knit init`.
- **`Docker` Installation**: Required for building and pushing images to the Knitfab platform.

**Optional:**

> - `kubectl` (recommended): Provides additional flexibility for debugging and interacting with Knitfab Kubernetes clusters.

Reviewer: In the example, `kubectl` is required to deploy ollama outside of Knitfab. This is not optional.

```
Once cloned, navigate to the `04.examples/news-classification` directory. You will find the following:
- **scripts:** This directory contains the essential files for training and deploying the model:
  - Python scripts (`train.py`, `evaluate.py`) for model training and evaluation.
  - `requirements.txt` files listing all required Python packages for dependency management.
  - `Dockerfile` used to build a Docker image of the application, enabling consistent deployment within the Knitfab Plan.
- **configs:** Contains the configuration for training and evaluation task.
- **plans:** Contains the Knitfab Plan YAML templates.

### Task
- [Step 1: Define the training and evaluation tasks.](#step-0-define-the-training-and-evaluation-tasks)
- [Step 2: Build and push docker image to Knitfab.](#step-1-build-and-push-docker-image-to-knitfab)
- [Step 3: Fine-tuning.](#step-2-fine-tuning)
- [Step 4: Set up Ollama for LLM-as-a-Judge evaluation.](#step-3-set-up-ollama-for-llm-as-a-judge-evaluation)
- [Step 5: LLM-as-a-judge evaluation.](#step-4-llm-as-a-judge-evaluation)
- [Step 6: Clean up.](#step-5-clean-up)

## Step 1: Define the training and evaluation tasks

> ### 1-1: Define Training Task

Reviewer: When using section numbers, they should be consistent format and numbering.

```suggestion
### 1.1. Define Training Task

```

... and, following items in this sections, should be "1.1.1", "1.1.2"... and so on.

> This section provides an in-depth explanation of the logic implemented in `train.py` for fine-tuning a news classification model. The script follows a structured pipeline that includes argument parsing, model preparation, dataset processing, training, and evaluation. Each component is designed to facilitate easy customization and optimization.

Reviewer: Refer where the code comes from.

```suggestion
This section provides an in-depth explanation of the logic implemented in `scripts/train/train.py` for fine-tuning a news classification model. The script follows a structured pipeline that includes argument parsing, model preparation, dataset processing, training, and evaluation. Each component is designed to facilitate easy customization and optimization.

```

```python
collator = DataCollatorWithPadding(tokenizer=self.tokenizer, padding="max_length")
```
- Ensures proper padding for sequences to maintain batch consistency.

```python
self.trainer = SFTTrainer(
    model=self.model,
    args=training_args,
    peft_config=peft_config,
    train_dataset=self.train_dataset,
    eval_dataset=self.eval_dataset,
    tokenizer=self.tokenizer,
    data_collator=collator,
)
```
- Initializes the `SFTTrainer` with the configured model, training arguments, dataset, and tokenizer.
- Combines all elements to execute the fine-tuning process efficiently.

### 5. Train and Evaluate the Model

The `run` method orchestrates the entire training process, from setting up the environment to saving the results.

#### Key Functionalities
- Loads datasets and initializes the trainer.
- Executes training and evaluation.
- Saves the trained model and evaluation metrics.

#### Code Breakdown
```python
self.trainer.train()
self.eval_results = self.trainer.evaluate()
self.save_results()
```
- Trains the model on the dataset.
- Evaluates performance on test data.
- Saves model checkpoints and metrics for later use.

> ### 1-2: LLM-as-a-judge Evaluation

Reviewer: ```suggestion
### 1.2. LLM-as-a-judge Evaluation

```

> This section presents an evaluation pipeline using an LLM to assess a fine-tuned news classification model. The `TestGPT2Model` class processes test cases, applies a category precision metric, and generates performance reports. This automated approach ensures consistent benchmarking and helps refine model accuracy.

Reviewer: ```suggestion
This section presents an evaluation pipeline using an LLM to assess a fine-tuned news classification model. The `TestGPT2Model` class in `scripts/evaluate/evaluate.py` processes test cases, applies a category precision metric, and generates performance reports. This automated approach ensures consistent benchmarking and helps refine model accuracy.

```

```
- Extract Run Outputs:
```bash
train_outputs=$(echo "$train_run" | jq -r '.[-1].outputs')
```
- Get Model Knit Id:
```bash
train_model_knit_id=$(echo "$train_outputs" | jq -r '.[0].knitId')
```

### 8. (Optional) Review the Run Log:

If you want to examine the logs generated during the training process, you can use the following commands:
- Get Run Id:
```bash
train_run_id=$(echo "$train_run" | jq -r '.[-1].runId')
```
- Show Run Log:
```bash
knit run show --log $train_run_id
```

### 9. (Optional) Download the Model:

If you need to access the trained model artifact directly, you can download it from the Knitfab platform using the following command:
```bash
knit data pull -x $train_model_knit_id ./out
```
This command downloads the trained model artifact from the Knitfab platform and stores it in the `./out` directory.

## Step 4: Set up Ollama for LLM-as-a-Judge evaluation.
This step deploys and configures Ollama, the application that will serve as the foundation for our LLM-as-a-judge evaluation process.
> [!Note]
> 
> If you plan to use a different LLM for evaluation, skip this step and configure your chosen LLM to ensure it's accessible during the evaluation Run.

> ### 1. Define Deployment and Service resources:

Reviewer: I understand that Ollama is needed for LLM-as-a-judge. But...

This section tells about Ollama and Kubernetes, not Knitfab.
Ollama *can* be anywhere reachable from Knitfab Run, and this example suggests deploy it in the same Kubernetes cluster and namespace for simplicity.

Make that point more clear.
Do not get user confused about "what is/is not Knitfab".

Create two YAML files, one for the Deployment and one for the Service, with the following content. Save these files, for example, as `ollama-deployment.yaml` and `ollama-service.yaml` respectively. These files define how your Ollama application will be deployed and accessed within your Kubernetes cluster.

> - `ollama-deployment.yaml`

Reviewer: These two yaml file can be conbined into one.

One file is easy to `kubectl apply -f`/`kubectl delete -f`.

```YAML
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ollama
  namespace: knitfab
  labels:
    app: ollama
spec:
  replicas: 1
  selector:
    matchLabels:
      app: ollama
  template:
    metadata:
      labels:
        app: ollama
    spec:
      containers:
        - name: ollama
          image: ollama/ollama
          resources:
            requests:
              cpu: "4"
              memory: "5Gi"
              nvidia.com/gpu: 1
            limits:
              cpu: "4"
              memory: "5Gi"
              nvidia.com/gpu: 1
          ports:
            - containerPort: 11434
          lifecycle:
            postStart:
              exec:
                command: ["sh", "-c", "ollama pull llama3.2"]
```

- `ollama-service.yaml`
```YAML
apiVersion: v1
kind: Service
metadata:
  name: ollama
  namespace: knitfab
spec:
  selector:
    app: ollama
  ports:
    - protocol: TCP
      port: 11434
      targetPort: 11434
  type: ClusterIP
```
For simplicity, in this example, we'll deploy the Ollama Deployment and Service into the same namespace as Knitfab.
### 2. Deploy Ollama application:
> [!Note]
> 

> > **Important Note Regarding GPU Access:** If you have only one GPU available in your Kubernetes cluster, you'll need to enable GPU sharing features like Time-Slicing or Multi-Instance GPU (MIG) to allow multiple Pods, including your Ollama Pod, to access it concurrently.  Without these features, your Ollama Deployment might fail or be unable to utilize the GPU.

Reviewer: ```suggestion
> **Important Note Regarding GPU Access:** If you have only one GPU available in your Kubernetes cluster, you may need to enable GPU sharing features like Time-Slicing or Multi-Instance GPU (MIG) to allow multiple Pods, including your Ollama Pod, to access it concurrently.  Without these features, your Ollama Deployment might fail or be unable to utilize the GPU.

```

If user have multiple GPU node, it cannot be any probrem. So, "will" is too strong, maybe.
